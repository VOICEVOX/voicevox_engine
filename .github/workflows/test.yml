name: test

on:
  push:
  pull_request:
    branches:
      - '**'

jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest]  # [ubuntu-latest, macos-latest, windows-latest]
        python: [3.7]
        include:
        - os: ubuntu-latest
          path: ~/.cache/pip
        # - os: macos-latest
        #   path: ~/Library/Caches/pip
        - os: windows-latest
          path: ~\AppData\Local\pip\Cache

    steps:
      - uses: actions/checkout@v2

      - name: Set up Python ${{ matrix.python }}
        uses: actions/setup-python@v2
        with:
          python-version: ${{ matrix.python }}

      - uses: actions/cache@v2
        id: pip-cache
        with:
          path: ${{ matrix.path }}
          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt', '**/requirements-test.txt', '**/requirements-dev.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-

      - name: Install libraries for ubuntu
        if: matrix.os == 'ubuntu-latest'
        run: sudo apt-get install libsndfile1-dev

      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip setuptools wheel
          pip install -r requirements-test.txt

      - run: pysen run lint

      - name: Run pytest and get coverage
        run: |
          coverage run --omit=test/* -m pytest

      - name: Submit coverage to Coveralls
        if: matrix.os == 'ubuntu-latest'
        run: coveralls --service=github
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Create coverage result
        if: github.event_name == 'pull_request' && matrix.os == 'ubuntu-latest'
        run: |
          coverage report > report.txt

      - name: Comment coverage result to Pull Requests
        if: github.event_name == 'pull_request' && matrix.os == 'ubuntu-latest'
        uses: actions/github-script@v5.0.0
        with:
          github-token: ${{ secrets.TOKEN_THAT_CAN_GITHUB_REPO_ACCESS }}
          script: |
            const fs = require('fs')
            const baseReport = fs.readFileSync('report.txt', 'utf8').toString().split('\n')
            let report = ''
            for (let i = 0; i < baseReport.length; i++) {
              const line = baseReport[i].split(' ').filter(v => v)
              if (i === 1 && line.length === 1) {
                report += "|:---|---:|---:|---:|\n"
              } else if (line.length === 1) {
                continue
              } else {
                if (i !== 0 && line.length === 4) {
                  const parcent = Number(line[3].replace("%", ""))
                  let color = 'green'
                  if (parcent < 50) {
                    color = 'red'
                  } else if (parcent < 90) {
                    color = 'orange'
                  }
                  line[3] = `![coverage-${parcent}%](https://img.shields.io/badge/coverage-${parcent}%25-${color}.svg)`
                }
                report += "|" + line.join("|") + "|\n"
              }
              if (line[0] === 'TOTAL') break
            }

            github.rest.issues.createComment({
              issue_number: context.issue.number,
              owner: context.repo.owner,
              repo: context.repo.repo,
              body: `## Coverage Result\n\n${report}`
            })

      - name: Check licenses
        shell: bash
        run: |
          python -m venv venv

          if [ -d "venv/Scripts" ]; then
            source venv/Scripts/activate
          else
            source venv/bin/activate
          fi

          pip install -r requirements-dev.txt
          python generate_licenses.py > /dev/null
