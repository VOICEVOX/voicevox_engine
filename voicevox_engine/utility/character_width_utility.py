"""文字列の全角・半角に関する utility"""

import re
from typing import NewType, TypeGuard

# 半角アルファベット文字列を示す型
HankakuAlphabet = NewType("HankakuAlphabet", str)


def is_hankaku_alphabet(text: str) -> TypeGuard[HankakuAlphabet]:
    """文字列が半角アルファベットのみで構成されているかを判定する"""
    return bool(re.fullmatch("[a-zA-Z]+", text))


def convert_zenkaku_alphabet_to_hankaku(text: str) -> str:
    """全角アルファベットを半角に変換する"""
    return text.translate(
        str.maketrans(
            "".join(chr(0xFF01 + i) for i in range(94)),
            "".join(chr(0x21 + i) for i in range(94)),
        )
    )
