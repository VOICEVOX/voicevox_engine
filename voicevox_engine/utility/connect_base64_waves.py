import base64
import io
from typing import List, Tuple

import numpy as np
import soundfile
from scipy.signal import resample


class ConnectBase64WavesException(Exception):
    def __init__(self, message: str):
        self.message = message


def decode_base64_waves(waves: List[str]) -> List[Tuple[np.ndarray, int]]:
    """
    base64エンコードされた複数のwavデータをデコードする
    Parameters
    ----------
    waves: list[str]
        base64エンコードされたwavデータのリスト
    Returns
    -------
    waves_nparray_sr: List[Tuple[np.ndarray, int]]
        (NumPy配列の音声波形データ, サンプリングレート) 形式のタプルのリスト
    """
    if len(waves) == 0:
        raise ConnectBase64WavesException("wavファイルが含まれていません")

    waves_nparray_sr = []
    for wave in waves:
        try:
            wav_bin = base64.standard_b64decode(wave)
        except ValueError:
            raise ConnectBase64WavesException("base64デコードに失敗しました")
        try:
            _data = soundfile.read(io.BytesIO(wav_bin))
        except Exception:
            raise ConnectBase64WavesException("wavファイルを読み込めませんでした")
        waves_nparray_sr.append(_data)

    return waves_nparray_sr


def connect_base64_waves(waves: List[str]) -> Tuple[np.ndarray, int]:
    waves_nparray_sr = decode_base64_waves(waves)

    max_sampling_rate = max([sr for _, sr in waves_nparray_sr])
    max_channels = max([x.ndim for x, _ in waves_nparray_sr])
    assert 0 < max_channels <= 2

    waves_nparray_list = []
    for nparray, sr in waves_nparray_sr:
        if sr != max_sampling_rate:
            nparray = resample(nparray, max_sampling_rate * len(nparray) // sr)
        if nparray.ndim < max_channels:
            nparray = np.array([nparray, nparray]).T
        waves_nparray_list.append(nparray)

    return np.concatenate(waves_nparray_list), max_sampling_rate
