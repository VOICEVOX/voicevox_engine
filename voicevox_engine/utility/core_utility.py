import os


def get_half_logical_cores() -> int:
    logical_cores = os.cpu_count()
    if logical_cores is None:
        return 0
    return logical_cores // 2
