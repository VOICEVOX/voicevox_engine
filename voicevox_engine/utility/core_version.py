from typing import Iterable

from packaging.version import Version
from packaging.version import parse as parse_version


def parse_core_version(version: str) -> Version:
    return parse_version(version)


def get_latest_core_version(versions: Iterable[str]) -> str:
    if len(versions) == 0:
        raise Exception('versions must be non-empty.')

    return str(
        max(map(parse_core_version, versions))
    )
