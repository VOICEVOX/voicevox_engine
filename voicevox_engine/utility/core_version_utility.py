from typing import Iterable

from semver.version import Version


def parse_core_version(version: str) -> Version:
    return Version.parse(version)


def get_latest_core_version(versions: Iterable[str]) -> str:
    if len(versions) == 0:
        raise Exception("versions must be non-empty.")

    return str(max(map(parse_core_version, versions)))
