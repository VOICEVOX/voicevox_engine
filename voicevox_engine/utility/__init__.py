from .connect_base64_waves import (
    ConnectBase64WavesException,
    connect_base64_waves,
    decode_base64_waves,
)
from .core_version_utility import get_latest_core_version, parse_core_version
from .mutex_utility import mutex_wrapper
from .path_utility import delete_file, engine_root, get_save_dir

__all__ = [
    "ConnectBase64WavesException",
    "connect_base64_waves",
    "decode_base64_waves",
    "get_latest_core_version",
    "parse_core_version",
    "delete_file",
    "engine_root",
    "get_save_dir",
    "mutex_wrapper",
]
