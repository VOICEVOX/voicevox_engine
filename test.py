import argparse
import wave
import numpy as np
from pathlib import Path
from voicevox_engine.core.core_initializer import initialize_cores
from voicevox_engine.tts_pipeline.tts_engine import make_tts_engines_from_cores, LATEST_VERSION
from voicevox_engine.utility.path_utility import engine_root, get_save_dir
from voicevox_engine.model import AudioQuery


# === CẤU HÌNH TRỰC TIẾP ===
HOST = ""
PORT = 0  # Không sử dụng API
USE_GPU = False
VOICEVOX_DIR          = Path("/home/trungquang/open_avt_phuc/OpenAvatarChat/src/handlers/tts/voicevox/voicevox_engine/voicevox_core")           # nếu có
VOICELIB_DIRS         = None
RUNTIME_DIRS          = [Path("/home/trungquang/open_avt_phuc/OpenAvatarChat/src/handlers/tts/voicevox/voicevox_engine/onnxruntime/onnxruntime-linux-x64-1.13.1/lib"),]
ENABLE_MOCK = False
INIT_PROCESSES = 2
LOAD_ALL_MODELS = True
CPU_NUM_THREADS = 4
DISABLE_MUTABLE_API = False
# ==========================



def synthesize_to_file(text: str, speaker: int, sample_rate: int, output_wav: Path):
    # Khởi tạo core và engine
    core_mgr = initialize_cores(
        use_gpu=USE_GPU,
        voicelib_dirs=VOICELIB_DIRS,
        voicevox_dir=VOICEVOX_DIR,
        runtime_dirs=RUNTIME_DIRS,
        cpu_num_threads=CPU_NUM_THREADS,
        enable_mock=ENABLE_MOCK,
        load_all_models=True,
    )
    tts_manager = make_tts_engines_from_cores(core_mgr)

    # Lấy engine phiên bản mới nhất
    engine = tts_manager.get_tts_engine(LATEST_VERSION)

    # Tạo query cho synthesis
    accent_phrases = engine.create_accent_phrases(text, speaker)
    query = AudioQuery(
        accent_phrases=accent_phrases,
        speedScale=1.0,
        pitchScale=0.0,
        intonationScale=1.0,
        volumeScale=1.0,
        prePhonemeLength=0.1,
        postPhonemeLength=0.1,
        pauseLength=None,
        pauseLengthScale=1.0,
        outputSamplingRate=sample_rate,
        outputStereo=False,
        kana="",
    )

    # Tổng hợp
    wav_array = engine.synthesize(query)
    # wav_array: numpy array shape (channels, frames)

    # Ghi file WAV
    samples = (wav_array * 32767).astype(np.int16)
    with wave.open(str(output_wav), 'wb') as wf:
        wf.setnchannels(samples.shape[0])
        wf.setsampwidth(2)
        wf.setframerate(sample_rate)
        wf.writeframes(samples.T.tobytes())


def main():
    parser = argparse.ArgumentParser(description="Một script đơn giản để tổng hợp giọng nói bằng VOICEVOX mà không cần API.")
    parser.add_argument('--text', type=str, help='Văn bản cần tổng hợp', default="不動産は")
    parser.add_argument('--speaker', type=int, help='ID của speaker', default=1)
    parser.add_argument('--sample_rate', type=int, help='Tỷ lệ mẫu', default=24000)
    parser.add_argument('--output', type=Path, help='Đường dẫn file WAV đầu ra', default="/home/trungquang/open_avt_phuc/OpenAvatarChat/src/handlers/tts/voicevox/output.wav")
    args = parser.parse_args()

    synthesize_to_file(
        text=args.text,
        speaker=args.speaker,
        sample_rate=args.sample_rate,
        output_wav=args.output,
    )
    print(f"Đã sinh file: {args.output}")


if __name__ == '__main__':
    main()
