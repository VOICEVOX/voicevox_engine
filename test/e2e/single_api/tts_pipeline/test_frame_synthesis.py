"""
/frame_synthesis API のテスト
"""

from test.utility import hash_wave_floats_from_wav_bytes

from fastapi.testclient import TestClient
from syrupy.assertion import SnapshotAssertion


def test_post_frame_synthesis_200(
    client: TestClient, snapshot: SnapshotAssertion
) -> None:
    query = {
        "f0": [
            0.0,
            0.0,
            46.63821792602539,
            46.63821792602539,
            46.63821792602539,
            46.63821792602539,
            46.63821792602539,
            46.63821792602539,
            46.63821792602539,
            46.63821792602539,
            46.396461486816406,
            46.396461486816406,
            46.61719512939453,
            46.66975021362305,
            83.099853515625,
            83.099853515625,
            82.96875762939453,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
        ],
        "volume": [
            0.0,
            0.0,
            0.3342486023902893,
            0.3342486023902893,
            0.3342486023902893,
            0.3342486023902893,
            0.3342486023902893,
            0.3342486023902893,
            0.3342486023902893,
            0.3342486023902893,
            0.12581685185432434,
            0.12581685185432434,
            0.316038578748703,
            0.36159414052963257,
            0.794084370136261,
            0.794084370136261,
            0.6428364515304565,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
        ],
        "phonemes": [
            {"phoneme": "pau", "frame_length": 2},
            {"phoneme": "t", "frame_length": 8},
            {"phoneme": "e", "frame_length": 2},
            {"phoneme": "s", "frame_length": 1},
            {"phoneme": "u", "frame_length": 1},
            {"phoneme": "t", "frame_length": 2},
            {"phoneme": "o", "frame_length": 1},
            {"phoneme": "pau", "frame_length": 10},
        ],
        "volumeScale": 1.0,
        "outputSamplingRate": 24000,
        "outputStereo": False,
    }
    response = client.post("/frame_synthesis", params={"speaker": 0}, json=query)
    assert response.status_code == 200

    # FileResponse 内の .wav から抽出された音声波形が一致する
    assert response.headers["content-type"] == "audio/wav"
    assert snapshot == hash_wave_floats_from_wav_bytes(response.read())
