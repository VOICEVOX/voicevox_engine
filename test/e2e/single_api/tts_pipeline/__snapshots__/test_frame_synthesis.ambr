# serializer version: 1
# name: test_post_frame_synthesis_200
  'MD5:1c385210acba238994604a8cee96aee3'
# ---
